### PR TITLE
feat(compiler): add support for source map generation

### LINKED ISSUE #14125: Generate source maps for compiled templates
The upcoming view engine (https://github.com/angular/angular/issues/14013) will rely on source maps for reporting the source location of errors.

For this we need a general support for source maps in our `OutputAst` and emitters (`TypeScriptEmitter`, `jitStatements`).

### PR DESCRIPTION
**Pull request by vicb**
fixes #14125

### CODE REVIEW

File: modules/@angular/compiler/src/output/abstract_emitter.ts [hunk 1]
```diff
@@ -77,36 +84,70 @@ export class EmitterVisitorContext {
...
+      // skip leading parts without source spans
+      while (spanIdx < spans.length && !spans[spanIdx]) {
+        col0 += parts[spanIdx].length;
+        spanIdx++;
+      }
+
+      while (spanIdx < spans.length) {
+        const span = spans[spanIdx];
+        const source = span.start.file;
+        const line = span.start.line;
```
Comment (vicb) [inline modules/@angular/compiler/src/output/abstract_emitter.ts]: srcLine, srcCol

File: modules/@angular/compiler/src/output/abstract_emitter.ts [hunk 2]
```diff
@@ -424,7 +466,7 @@ export function escapeIdentifier(
 function _createIndent(count: number): string {
   let res = '';
   for (let i = 0; i < count; i++) {
-    res += '  ';
+    res += _INDENT_WITH;
```
Comment (vicb) [inline modules/@angular/compiler/src/output/abstract_emitter.ts]: width

File: modules/@angular/compiler/src/output/output_jit.ts [hunk 3]
```diff
@@ -7,7 +7,6 @@
  */
```
Comment (vicb) [inline modules/@angular/compiler/src/output/output_jit.ts]: add src map

File: modules/@angular/compiler/src/output/js_emitter.ts [hunk 4]
```diff
@@ -18,18 +18,27 @@ import {ImportResolver} from './path_util';
...
 export class JavaScriptEmitter implements OutputEmitter {
   constructor(private _importResolver: ImportResolver) {}
+
   emitStatements(genFilePath: string, stmts: o.Statement[], exportedVars: string[]): string {
     const converter = new JsEmitterVisitor(genFilePath, this._importResolver);
     const ctx = EmitterVisitorContext.createRoot(exportedVars);
     converter.visitAllStatements(stmts, ctx);
+
     const srcParts: string[] = [];
     converter.importsWithPrefixes.forEach((prefix, importedFilePath) => {
```
Comment (vicb) [inline modules/@angular/compiler/src/output/js_emitter.ts]: handle import shift

File: modules/@angular/compiler/src/output/ts_emitter.ts [hunk 5]
```diff
@@ -41,26 +41,40 @@ export function debugOutputAstAsTypeScript(ast: o.Statement | o.Expression | o.T
...
     const srcParts: string[] = [];
+
     converter.reexports.forEach((reexports, exportedFilePath) => {
       const reexportsCode =
           reexports.map(reexport => `${reexport.name} as ${reexport.as}`).join(',');
       srcParts.push(
           `export {${reexportsCode}} from '${this._importResolver.fileNameToModuleName(exportedFilePath, genFilePath)}';`);
     });
+
     converter.importsWithPrefixes.forEach((prefix, importedFilePath) => {
```
Comment (vicb) [inline modules/@angular/compiler/src/output/ts_emitter.ts]: handle

File: modules/@angular/compiler/test/output/abstract_emitter_spec.ts [hunk 6]
```diff
@@ -32,5 +37,126 @@ export function main() {
...
+        ctx.print(createSourceSpan(fileA, 1), 'o1');
+        ctx.print(createSourceSpan(fileB, 0), 'o2');
+        ctx.print(createSourceSpan(fileB, 1), 'o3');
+        const sm = ctx.toSourceMapGenerator('o.js').toJSON();
+        expect(sm.sources).toEqual([fileA.url, fileB.url]);
+        expect(sm.sourcesContent).toEqual([fileA.content, fileB.content]);
+      });
+
+      it('should generate a correct mapping', () => {
+        ctx.print(createSourceSpan(fileA, 0), 'fileA-0');
```
Comment (vicb) [inline modules/@angular/compiler/test/output/abstract_emitter_spec.ts]: Width2

File: modules/@angular/compiler/test/output/abstract_emitter_spec.ts [hunk 7]
```diff
@@ -32,5 +37,126 @@ export function main() {
          () => { expect(escapeIdentifier('class', false, false)).toEqual('class'); });
     });
 
+    describe('EmitterVisitorContext', () => {
```
Comment (vicb) [inline modules/@angular/compiler/test/output/abstract_emitter_spec.ts]: run tests in CJS only

**Review by tbosch — APPROVED** [review]

**Review by mhevery — APPROVED** [review]

**Pull request closed without merging**